Ethan's revised function for computing data extents from L3_mosaics.csv and saving as .geojson and .shp

ekcomputer opened the pull request:
Also modified mair_geoviews: When viewing in geoviews, I made the active tools pan and zoom by default. Also added Open Streetmap basemap- maybe you don't want to enable this by default.

rocheseb commented on notebooks/mair_geoviews.ipynb:
```diff
@@ -1013,7 +498,7 @@
     "    alpha  = 1             # opacity of the plotted field, 0 = completely transparant, 1 = opaque\n",
     "    \n",
-    "    return show_map(lon, lat, z, title=title, width=width, height=height, clim=clim, cmap=cmap, alpha=alpha)\n",
+    "    return show_map(lon, lat, z, title=title, width=width, height=height, clim=clim, cmap=cmap, alpha=alpha, background_tile=OSM)\n",
```
Seems I hadn't updated the notebook along with the latest mair_geoviews updates so the show_map function in the notebook is out of sync with the one in mair_geoviews

rocheseb commented on msatutil/make_spatial_index.py:
```diff
@@ -0,0 +1,139 @@
+    import pyogrio
+    engine = 'pyogrio'
+except:
+    engine = 'fiona'
```
We should update the dependencies in setup.py instead of having this try/except

rocheseb commented on msatutil/make_spatial_index.py:
```diff
@@ -0,0 +1,139 @@
+
+def validDataArea2Gdf(ds, simplify=None):
+    '''
+    This function converts valid data areas from a msatutil.msat_dset.msat_dset (Netcdf4 Dataset) into a MultiPolygon Shapefly geometry. It takes about an hour to run for 1,000 mosaics.
```
do you know which line is time consuming?

ekcomputer replied:
I think the slowest parts are saving intermittently and  loading the data from the bucket, even if running on the cloud (and especially for L2 data). At least the saving frequency can be adjusted as a function argument.

ekcomputer replied:
As currently written, this script will not be practical for the 75k L2 files listed, especially since the intermediate saving will keep writing a larger and larger file as the script goes on, but it should work fine over smaller quantities.

rocheseb commented on msatutil/make_spatial_index.py:
```diff
@@ -0,0 +1,139 @@
+    return multipolygon
+
+
+if __name__ == '__main__':
```
can you put the contents into a `main` function, and only call `main()` under `if __name__ == '__main__':`?

the main() function would parse arguments with `argparse` and then call a function that does everything you are doing here.
This function arguments would be e.g.

`def save_geojson(L3_mosaics_catalogue_pth, load_from_chkpt=True,simplify_tol=None,save_frequency=50, ...)`

There shouldn't be any inputs needing to be modified in the code.

rocheseb commented on msatutil/make_spatial_index.py:
```diff
@@ -0,0 +1,139 @@
+            df = gpd.read_file(catalogue_shp_out_pth, engine=engine)
+        else:
+            df = pd.read_csv(L3_mosaics_catalogue_pth)            
+        df = df[:3]  # Testing
```
does this need to stay?

rocheseb commented on msatutil/make_spatial_index.py:
```diff
@@ -0,0 +1,139 @@
+        df = df[:3]  # Testing
+    else:  # on GCS 'Linux-6.2.0-1013-gcp-x86_64-with-glibc2.35'
+        working_dir = '~/msat_spatial_idx'
+        catalogue_shp_out_pth = os.path.join(working_dir, f'L3_mosaics_20240208{tol_str}.shp')
+        if load_from_chkpt and os.path.exists(os.path.expanduser(catalogue_shp_out_pth)):
+            df = gpd.read_file(catalogue_shp_out_pth, engine=engine)
+        else:
+            df = pd.read_csv(L3_mosaics_catalogue_pth,
+                            storage_options={'token': 'cloud'})
```
can you pull this out of the if statement so it doesn't need to be repeated?

ekcomputer replied:
I rewrote this section without all the if statements.

ekcomputer commented:
Okay, ready for review! I made the changes you highlighted, and also added a notebook demoing the make_spatial_index function. Added functionality to run on L2 files.

rocheseb commented:
@ekcomputer before merging can you add how to run the new code [here in the readme](https://github.com/rocheseb/msatutil/blob/master/README.md?plain=1#L55) with a `msatutil.make_spatial_index` section?

ekcomputer commented:
Sorry for the delay! I have made these README changes, plus a few other bug fixes, and integration with mair_ls to only process the latest file versions. Unfortunately, the CSV files you made now refer to old cloud paths that got deleted. It would be helpful if you and Marcus eventually regenerated the L2pp/L3 segments/L3 mosaics csvs.

rocheseb commented:
@ekcomputer thanks I'll update the version and merge. For the CSVs of file paths I now generate my own, maybe I'll add the code to do that in the repo
